src/services/VerifactuService.php

```diff
@@ -129,52 +147,80 @@ public static function registerInvoice(InvoiceSubmission $invoice)
     /**
      * Cancels an invoice with AEAT via VERI*FACTU.
      *
-     * @param InvoiceCancellation $cancellation
      * @return InvoiceResponse
      */
     public static function cancelInvoice(InvoiceCancellation $cancellation)
     {
-        $validation = $cancellation->validate();
+        // 1. Validate input (excluding hash which will be generated)
+        $validation = $cancellation->validateExceptHash();
+
         if ($validation !== true) {
             throw new \InvalidArgumentException('InvoiceCancellation validation failed: ' . print_r($validation, true));
         }
+
+        // 2. Generate hash (huella)
         $cancellation->hash = HashGeneratorService::generate($cancellation);
+
+        // 3. Final validation including hash
+        $finalValidation = $cancellation->validate();
+
+        if ($finalValidation !== true) {
+            throw new \InvalidArgumentException('InvoiceCancellation final validation failed: ' . print_r($finalValidation, true));
+        }
         $xml = self::buildCancellationXml($cancellation);
-        $signedXml = XmlSignerService::signXml(
-            $xml,
-            self::getConfig(self::CERT_PATH_KEY),
-            self::getConfig(self::CERT_PASSWORD_KEY)
-        );
+        $useContent = false;
+        $env = self::$config['environment'] ?? null;
```

Comment:
We cannot let library behave different on development environment. Library must work always the same in prod and sandbox, with the only thins to be changed must be configuration, like endpoint urls or file paths, but we cannot change how code behaves or also introducing custom constants for development use only, like CERT_CONTENT_KEY.